common/thorhelper/thorstore.cpp

```diff
@@ -0,0 +1,286 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "jliball.hpp"
+#include "jsocket.hpp"
+#include "thorfile.hpp"
+
+#include "eclhelper.hpp"
+#include "eclrtl.hpp"
+#include "eclrtl_imp.hpp"
+
+#include "dautils.hpp"
+#include "dadfs.hpp"
+
+#include "thorstore.hpp"
+
+
+//If the path associated with a plane contains one or more hashes in a row those hashes
+//are replaced with the device number, padded with 0s to the width of the number of hashes
+StringBuffer & expandPlanePath(StringBuffer & target, const char * path, unsigned device)
+{
+    for (;;)
+    {
+        const char * hash = strchr(path, '#');
+        if (!hash)
+            break;
+
+        target.append(hash-path, path);
+        unsigned width = 1;
+        while (hash[1] == '#')
+        {
+            hash++;
+            width++;
+        }
+
+        target.appendf("%0*u", width, device);
+        path = hash+1;
+    }
+    target.append(path);
+    return target;
+}
+
+CStorageHostGroup::CStorageHostGroup(const IPropertyTree * _xml)
+: xml(_xml)
+{
+}
+
+const char * CStorageHostGroup::queryName() const
+{
+    return xml->queryProp("@name");
+}
+
+const char * CStorageHostGroup::queryHost(unsigned idx) const
+{
+    VStringBuffer xpath("hosts[%u]", idx+1);
+    return xml->queryProp(xpath);
+}
+
+bool CStorageHostGroup::isLocal(unsigned device) const
+{
+    const char * hostname = queryHost(device);
+    if (hostname)
+    {
+        //MORE: Likely to be inefficient - should search differently?
+        IpAddress ip(hostname);
+        return ip.isLocal();
+    }
+    return false;
+}
+
+/*
+ * A StoragePlane represents a set of storage devices that are used together to store a distributed file.  Each
+ * storage plane has a fixed number of logical devices.  If there are > 1 then the storage plane index can be included
+ * within the file path - which allows data to be striped across multiple devices.
+ *
```

Comment:
The storage plane index is really interesting. Would it make sense to allow for aliases to be configured for these different storage plane indexes in the future? IE: Would something like S3_INSURANCE://car_data make things more portable than something like S3://1/car_data? (Making assumptions on how the URL is structured.)

Reply:
There is a difference between the storage plane name and the syntax used to access a physical file.  The ecl users will never need to use the s3:// syntax.  They only need to use the name of the storage plane.
You could also create a storage plane for one of he specific instances,  but it won't be very useful because it will only contain a subsection of the physical file parts.